internal/pipeline/executor.go

```diff
@@ -281,3 +313,46 @@ func fnChain(pl *kptfilev1alpha2.Pipeline, pkgPath pkg.UniquePath) ([]kio.Filter
 	}
 	return runners, nil
 }
+
+// trackInputFiles records file paths of input resources in the hydration context.
+func trackInputFiles(hctx *hydrationContext, input []*yaml.RNode) error {
+	if hctx.inputFiles == nil {
+		hctx.inputFiles = sets.String{}
+	}
+	for _, r := range input {
+		path, _, err := kioutil.GetFileAnnotations(r)
+		if err != nil {
+			return fmt.Errorf("path annotation missing: %w", err)
+		}
+		hctx.inputFiles.Insert(path)
+	}
+	return nil
+}
+
+// trackOutputfiles records the file paths of output resources in the hydration
+// context. It should be invoked post hydration.
+func trackOutputFiles(hctx *hydrationContext) error {
+	outputSet := sets.String{}
+
+	for _, r := range hctx.root.resources {
+		path, _, err := kioutil.GetFileAnnotations(r)
+		if err != nil {
+			return fmt.Errorf("path annotation missing: %w", err)
+		}
+		outputSet.Insert(path)
+	}
+	hctx.outputFiles = outputSet
+	return nil
+}
+
+// pruneResources compares the input and output of the hydration and prunes
+// resources that are no longer present in the output of the hydration.
+func pruneResources(hctx *hydrationContext) error {
```

Comment:
presumably, this functionality exists in current `kpt fn run`. What's the plan for converging this? We need pruning for imperative functions as well.

Reply:
One thing to note is that during hydration, we read resources as we traverse the pkg hierarchy (and also gather resources by applying pipeline) and write it finally in the end. The read/write flow is very different from imperative run or just pkg level operation. For pruning, you want to track resources at reading time and writing time, so the flow will be different for hydration vs imperative workflow.

For ex, imperative runs etc. uses `kio.LocalPkgReaderWriter` https://github.com/kubernetes-sigs/kustomize/blob/kyaml/v0.10.15/kyaml/kio/pkgio_reader.go#L116 where it tracks the file while `.Read()` and prunes files while `Write(resources)` call. So that is useful at the pkg level, but isn't suited to be used for the hydration workflow.

What do you have in mind when you think of `converging` ?

Reply:
I'm not following why declarative and imperative are different when it comes to package io and prunning. There are 3 high level stages:

1.  Get the current set resources across all subpackages. This can be done in different ways. a) Aggregate `inputResources` as part of recursive algorithm b) `Read()` on top-level package with `includeSubpackages`
2. Runs functions as part recursive hydration to calculate the resultant resources across all packages. This is the `outputResources`.
3. Write the `outputResources` to disk across all packages with pruning (diff step 1 and 3).

The only difference between imperative and declarative is step 2, no? Pruning is a concern for package reader/writer (steps 1 and 3), and not the hydration process.  It would be weird to duplicate the logic in step 3.

Reply:
For pruning to work, we need to track the `input` resources (A) and `output` resources (B) so that we can compute the difference `A-B` at the end and perform the deletion from FS. Keeping track of `input` as we read resources during `imperative` or `declarative flow` is where we need the right abstraction so that we can de-dupe the logic for pruning. One way to do it will be have some sort of `resourceTracker` that you can provide to the abstractions `pkg.LocalReaderWriter()` or `hydrationContext` so that these methods could call `resourceTracker.Track(record)` as they read resource or feed to the function chains and at the end have some sort of `resourceTracker.Reconcile(output)` to get the difference between A-B and do the pruning.

Would like to have a chat about this. Scheduling 30 mins tomorrow to go over this.

Reply:
Can you file an issue to follow up on this. I'm still missing something given current implementation of declarative flow dooesn't reimplement prunning.

Reply:
Filed an issue here https://github.com/GoogleContainerTools/kpt/issues/1609 to track this conversation.